### PR TITLE
Fix routing infinite loop, bad scoring

### PR DESCRIPTION
**Pull request by rustyrussell**
Three small fixes: the scoring ones may have a large effect in real routing scenarios!

Fixes: https://github.com/ElementsProject/lightning/issues/7092

### CODE REVIEW

**Review by vincenzopalazzo — APPROVED** [review]
ACK https://github.com/ElementsProject/lightning/pull/7127/commits/4b06cde5780d9cb40d7b5ae362350d2dbf8b93b4

File: plugins/libplugin-pay.c [hunk 1]
```diff
@@ -773,14 +773,15 @@ static u64 route_score(u32 distance,
...
 
 	/* Smoothed harmonic mean to avoid division by 0 */
-	u64 costs = (cmsat * rmsat * bias) / (cmsat + rmsat + bias + 1);
+	double costs = (cmsat * rmsat * bias) / (cmsat + rmsat + bias + 1);
 
 	if (costs > 0xFFFFFFFF)
-		costs = 0xFFFFFFFF;
-	return costs;
+		return 0xFFFFFFFF;
+	return (u64)costs;
```
[the remark this answers is not in the document]
Comment (rustyrussell) [reply plugins/libplugin-pay.c:784]: It's just to be explicit that we're changing types: the type changes were kind of hidden before.

**Comment by rustyrussell** [comment]
The final commit is wrong: we do, indeed, sum the inputs to that function.  However, we shouldn't.  Will fix separately (now I'm writing tests!)

**Comment by rustyrussell** [comment]
Even after fixing the capacity bias, we had a similar problem with our risk values (being 0).  In fact, the scoring callback was counterintuitive, so I've fixed that entirely.

The main contribution was to spend a day writing a proper unit test of the routing code, which uncovered this confusion and now gives me much more optimism that this is correct.

Please re-review: the first two commits are the same, though.

**Pull request merged**